File: backend/transcendence/game/game_logic/Paddle.py
from datetime import datetime
from .const_vars import *

class Paddle:
	def __init__(self, side):
		self.screen_height = SCREEN_HEIGHT
		self.screen_width = SCREEN_WIDTH
		self.height = 50
		self.width = 4
		self.wall_distance = 10
		self.half_height = self.height / 2
		self.half_width = self.width / 2
		self.step = 6
		self.side = side
		self.y = self.screen_height / 2
		self.x = self.__get_x(side)
		self.state = None
		self.last_time = None

	def get_position(self):
		return self.y - self.half_height

	def update(self):
		if self.state:
			if not self.last_time:
				self.last_time = datetime.now()
				value = 1
			else:
				current_time = datetime.now()
				value = (current_time.timestamp() - self.last_time.timestamp()) * 1000
				self.last_time = current_time
			step = value * PADDLE_SPEED
			if self.state == "up":
				self.y -= step
			else:
				self.y += step
			if self.y > self.screen_height - self.half_height:			
				self.y = self.screen_height - self.half_height
			elif self.y < self.half_height:
				self.y = self.half_height

	def set_state(self, key, status):
		if self.state == key or not self.state:
			if status == "pressed":
				self.state = key
			else:
				self.state = None
				self.last_time = None

	def get_colision_point(self, old_ball_x, old_ball_y, new_ball_x, new_ball_y, ball_radius):
		x_colision = False
		y_colision = False
		if self.side == "left":
			x_colision = abs(new_ball_x - self.x) <= (ball_radius + self.half_width)
			y_colision = abs(new_ball_y - self.y) <= (ball_radius + self.half_height)
			new_x = self.x + ball_radius
		elif self.side == "right":
			x_colision = abs(self.x - new_ball_x) <= (ball_radius + self.half_width)
			y_colision = abs(self.y - new_ball_y) <= (ball_radius + self.half_height)
			new_x = self.x - ball_radius
		if x_colision and y_colision:
			m = (new_ball_y - old_ball_y) / (new_ball_x - old_ball_x)
			b = new_ball_y - m * new_ball_x
			new_y = m * new_x + b	
		else:
			return None
		hit_paddle_value = self.y - self.half_height - new_y
		hit_paddle_percentage = abs(hit_paddle_value * 100 / self.height)
		return {"x": new_x, "y": new_y, "hit_paddle_percentage": hit_paddle_percentage}

	def end_game_position(self):
		self.y = self.screen_height / 2

	def __get_x(self, side):
		if side == "left":
			return self.wall_distance + self.width
		return self.screen_width - self.wall_distance - self.width

File: srcs/frontend/app/game/Paddle.js
import * as const_vars from "./const_vars.js";
import get_position_percentage from "./logicUtils.js"

export class Paddle {
	constructor(side) {
		this.screen_height = const_vars.SCREEN_HEIGHT;
		this.screen_width = const_vars.SCREEN_WIDTH;
		this.height = const_vars.PADDLE_HEIGHT;
		this.width = const_vars.PADDLE_WIDTH;
		this.wall_distance = 10;
		this.half_height = this.height / 2;
		this.half_width = this.width / 2;
		this.step = 6;
		this.side = side;
		this.y = this.screen_height / 2;
		this.x = this.#getX(side);
		this.state = null;
		this.last_time = null;
	}

	getPosition() {
		return get_position_percentage(this.y - this.half_height, const_vars.SCREEN_HEIGHT);
	}

	update() {
		if (this.state) {
			let value;
			if (!this.last_time) {
				this.last_time = Date.now();
				value = 1;
			}
			else {
				const current_time = Date.now();
				value = (current_time - this.last_time);
				this.last_time = current_time;
			}
			this.step = value * const_vars.PADDLE_SPEED;

			if (this.state === 'up')
				this.y -= this.step;
			else if (this.state === 'down')
				this.y += this.step;

			if (this.y > this.screen_height - this.half_height)
				this.y = this.screen_height - this.half_height;
			else if (this.y < this.half_height)
				this.y = this.half_height;
		}
	}

	setState(key, status) {
		if (this.state === key || !this.state) {
			if (status === 'pressed')
				this.state = key;
			else {
				this.state = null;
				this.last_time = null;
			}
		}
	}

	getColisionPoint(oldBallX, oldBallY, newBallX, newBallY, ballRadius) {
		let xColision = false;
		let yColision = false;
		let newX, newY;

		if (this.side === 'left') {
			xColision = Math.abs(newBallX - this.x) <= (ballRadius + this.half_width);
			yColision = Math.abs(newBallY - this.y) <= (ballRadius + this.half_height);
			newX = this.x + ballRadius;
		}
		else if (this.side === 'right') {
			xColision = Math.abs(this.x - newBallX) <= (ballRadius + this.half_width);
			yColision = Math.abs(this.y - newBallY) <= (ballRadius + this.half_height);
			newX = this.x - ballRadius;
		}

		if (xColision && yColision) {
			const m = (newBallY - oldBallY) / (newBallX - oldBallX);
			const b = newBallY - m * newBallX;
			newY = m * newX + b;
		}
		else
			return null;

		const hitPaddleValue = this.y - this.half_height - newY;
		const hitPaddlePercentage = Math.abs((hitPaddleValue * 100) / this.height);

		return { x: newX, y: newY, hit_paddle_percentage: hitPaddlePercentage };
	}

	endGamePosition() {
		this.y = this.screen_height / 2;
	}

	#getX(side) {
		if (side === 'left')
			return this.wall_distance + this.width;
		else
			return this.screen_width - this.wall_distance - this.width;
	}
}
